### ISSUE TITLE
Performance issue with Requests.php decompress() / compatible_gzinflate() functions

### ISSUE THREAD

**Issue by adam-pro** [rmccue/Requests#153]
Currently the function: decompress() speculatively attempts the following in turn:

1. gzdecode(...)
2. gzinflate(...)
3. self::compatible_gzinflate(...)
4. gzuncompress(...)

However, #3 compatible_gzinflate(...) includes within it (at Lines 834 to 838) a "fallback" call to gzinflate(...) which is identical to that attempted at #2.

Hence if #2 fails. it may be uselessly repeated within #3.

One solution may be to do BOTH the following:
* MOVE the "fallback" in compatible_gzinflate(...) to the beginning of that function (i.e. make it the first inflation attempted), and 
* REMOVE from decompress() function the gzinflate(...) attempt at #2.

**Comment by ozh** [comment]
You should submit one pull request with all your proposed changes, and explain them, instead of filing a dozen separated issues... Just sayin' :)

**Comment by adam-pro** [comment]
I'm new to github, but in my experience bugs/issues should be logged at an atomic level so they can be easily assessed / prioritised / managed.

In any event these aren't pull requests as I haven't actually made any code changes, I'm just highlighting them so the software author who knows his code can assess whether they have merit.

**Comment by ozh** [comment]
Yep, you're right about keeping things on an atomic level. The beauty of pull requests, if done properly, is that you can "cherry pick" individual changes ("commits"). It also saves the author lot of time and increase vastly the chances that he'll actually test your proposed changes :)
